### PR TITLE
[MERGE][IMP] website_slides: refactor content management + vimeo support

### PR DESCRIPTION
**Pull request by awa-odoo**
This merge commit has five sub-sections:

1. Rename 'slide_type' and 'datas' fields

To 'slide_category' and 'binary_content', in order to ease the following
changes.

See underlying commits for detailed information about those points.

2. Completely refactor the way we handle external content in the course slides.

This includes both technical (new fields, ...) and functional (forms) changes.
With these changes, we aim to make it much clearer for the end user on how to
introduce new content to its course, both for local files and external links.

3. Introduce support for Vimeo

Vimeo is a platform for professionals that manage video content, similarly to
YouTube.
The changes include allowing to introduce Vimeo links in your slides, correctly
embed the video, automatically set it as completed in the last 30 seconds,
fetch video metadata to autocomplete fields when URL is input, ...

4. Slightly improve e-learning views

See underlying commits for detailed information about those points.

5. Rename 'webpage' to 'article'

Indeed, 'webpage' was confusing because it was leading you to believe that you
would have to link external content from another website.

See underlying commits for detailed information about those points.

LINKS

Task-2510174
UPG PR odoo/upgrade#2498

### CODE REVIEW

File: addons/website_slides/models/slide_slide.py [hunk 1]
```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
...
+
+            slide.embed_code = embed_code
+
+    @api.depends('video_url')
+    def _compute_video_source_type(self):
+        for slide in self:
+            video_source_type = False
+            youtube_match = re.match(self.YOUTUBE_VIDEO_ID_REGEX, slide.video_url) if slide.video_url else False
+            if youtube_match:
+                if youtube_match and len(youtube_match.groups()) == 2 and len(youtube_match.group(2)) == 11:
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:402]: Mmmh ? would maybe extract to a sub tool method, to ease method understanding, like
```
if self._slide_video_is_youtube(self.video_url):
    video_source_type = 'youtube'
 elif self._slide_video_is_google(self.video_url):
    video_source_type = 'google_drive'
```

File: addons/website_slides/models/slide_slide.py [hunk 2]
```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
...
                 else:
-                    # embed google doc video
-                    record.embed_code = '<iframe src="//drive.google.com/file/d/%s/preview" allowFullScreen="true" frameborder="0"></iframe>' % (record.document_id)
+                    slide.video_youtube_id = False
             else:
-                record.embed_code = False
+                slide.video_youtube_id = False
 
-    @api.onchange('url')
+    @api.depends('slide_type', 'document_url', 'video_url', 'video_source_type')
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:434]: Somehow I think having a unique stored url field could help as it would depend on external_url only. Would also allow to remove dependency on video_source_type itself, as in the end it is not really important for google drive id computation if I understand well.

File: addons/website_slides/models/slide_slide.py [hunk 3]
```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
...
+            slide_metadata = self._fetch_external_metadata()
+            if slide_metadata:
+                self.update(slide_metadata)
+
+    @api.onchange('document_data_pdf')
+    def _on_change_document_data_pdf(self):
+        if self.slide_type == 'document' and self.source_type == 'local_file' and self.document_data_pdf:
+            completion_time = self._get_completion_time_pdf(base64.b64decode(self.document_data_pdf))
+            if completion_time:
+                self.completion_time = completion_time
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:485]: Shouldn't it be computed / editable ? based on either datas for pdf, youtube information, ... ?

File: addons/website_slides/models/slide_slide.py [hunk 4]
```diff
@@ -335,58 +361,116 @@ def _compute_user_membership_id(self):
...
+
+    @api.onchange('image_data')
+    def _on_change_image_data(self):
+        if self.slide_type == 'infographic' and self.source_type == 'local_file' and self.image_data:
+            self.image_1920 = self.image_data
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:490]: Why not a computed field ? Would allow to have this behavior implemented in all cases (form view, import, frontend, ...)

**Review by tde-banana-odoo — COMMENTED** [review]
Validation: models, fields. A bit "on hold" as it is probably better to discuss it de vive voix and see if we keep current implementation or go deeper in trying to clean.

File: addons/website_slides/static/src/js/slides_course_fullscreen_player.js [hunk 5]
```diff
@@ -129,6 +130,98 @@
...
+    /**
+     * This widget is responsible of loading the Vimeo video.
+     *
+     * Similarly to the YouTube implementation, the widget will trigger an event `change_slide` when
+     * the video is at its end, and `slide_completed` when the player is at 30 sec before the end of
+     * the video (30 sec before is considered as completed).
+     *
+     * See https://developer.vimeo.com/player/sdk/reference for all the API documentation.
+     */
+    var VideoPlayerVimeo = publicWidget.Widget.extend({
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_course_fullscreen_player.js:142]: Side linting question: isn't it better to have it in its own small file ? just to ease review / reading ?

**Review by awa-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 

I'm finally done renaming / adapting everything following our big brainstorming.

Also took a look at your first validation comments and addressed them.
A lot has changed since then so I might have missed something, a few comments are not relevant anymore after the re-refactoring.

Side note, since I pushed the refactoring before starting to look at the detailed comment, I can't make a nice "answer review".

So instead I'm forced to answer to some of them in this review global comment.
I will quote each comment and add my answer below.
It's not going to be very convenient for you, sorry about that:

> Just to be sure: why do we consider this to be an error here and not at model level (aka no constraint) ?
> Just to help people when they upload their content (adding it twice) ?

(Side note: this was about having an error message when uploading twice the same video only on the frontend).
I admit I re-implemented the previous behavior without questioning it.
I'm guessing we consider that people on the frontend need a little more help than in the backend where "you know what you are doing".
I agree it's absolutely not true, maybe we should make it a constraint, not sure how we would handle migration though.

> What do we receive ? Shouldn't we whitelist (aka url and channel_id) ? Or do we receive all uploaded slide values, in which case we could rename data into slide_values ? And why is slide_type separated :D ?
> Final conclusion: we could have slide_type, url, channel_id, **slide_values

(Side note: this was about the slide addition controller method).
I ended up adding the parameters in the arguments, it seems we never receive anything fancy.

> Shouldn't it be computed / editable ? based on either datas for pdf, youtube information, ... ?

(Side note: this was about the "completion_time" field)
From what I understand, we compute the completion_time in the frontend form only and in the preview on the website
Otherwise we want to avoid it, since it's a "nice to have magic".

> Why not a computed field ? Would allow to have this behavior implemented in all cases (form view, import, frontend, ...)

I made 'image_1920' a computed field, seems to work fine, good idea!

==================================

For the rest of the answers I was able to put them in my review where they belong.

I will re-launch a full testing with Luc since we changed a LOT of stuff.
But I wanted to answer to you before we re-do a validation loop.

Thanks for the validation and also for your help in our "let's make website_slides great again" brainstorming session.

Cheers!

File: addons/website_slides/models/slide_slide.py [hunk 6]
```diff
@@ -315,62 +383,96 @@ def _compute_user_membership_id(self):
...
-                try:
-                    pdf.getNumPages()
-                except PyPDF2.utils.PdfReadError:
-                    return
-                self.completion_time = (5 * len(pdf.pages)) / 60
-            else:
-                self.slide_category = 'infographic'
-                self.image_1920 = self.binary_content
-                self.binary_content = None
+        if self.url or self.document_google_url or self.image_google_url or self.video_url:
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:497]: Isn't it the same ? (url -> other are related)

File: addons/website_slides/models/slide_slide.py [hunk 7]
```diff
@@ -315,62 +383,96 @@ def _compute_user_membership_id(self):
...
+        if self.slide_category == 'document' and self.source_type == 'local_file' and self.document_binary_content:
+            completion_time = self._get_completion_time_pdf(base64.b64decode(self.document_binary_content))
+            if completion_time:
+                self.completion_time = completion_time
+
+    @api.onchange('slide_category')
+    def _on_change_slide_category(self):
+        """ Prevents mis-match when ones uploads an image and then a pdf without saving the form. """
+        if self.slide_category != 'infographic' and self.image_binary_content:
+            self.image_binary_content = False
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:517]: As those are related, is it necessary ? Shoudln't we empty binary_content directly when leaving infographic or document category ?

File: addons/website_slides/models/slide_slide.py [hunk 8]
```diff
@@ -401,29 +503,29 @@ def create(self, values):
             # 'website_published' is handled by mixin
             values['date_published'] = False
 
-        if values.get('slide_category') == 'infographic' and not values.get('image_1920'):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py]: Ok, understanding the compute on image then #smart

File: addons/website_slides/models/slide_slide.py [hunk 9]
```diff
@@ -401,29 +503,29 @@ def create(self, values):
...
-            for key, value in doc_data.items():
-                values.setdefault(key, value)
 
         slide = super(Slide, self).create(values)
 
+        # avoid fetching external metadata when installing the module (i.e. for demo data)
+        # we also support a context key if you don't want to fetch the metadata when creating a slide
+        if any(values.get(url_param) for url_param in ['url', 'video_url', 'document_google_url', 'image_google_url']) \
+           and not self.env.context.get('install_mode', False) \
+           and self.env.context.get('website_slides_fetch_metadata', True):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:562]: Debate ! ``website_slides_fetch_metadata, True`` or ``website_slides_skip_fetch_metadata, False`` ? I generally use the negative term (allowing to do ``if not self.env.context.get('website_slides_skip_fetch_metadata')`` without default). Fight !

File: addons/website_slides/models/slide_slide.py [hunk 10]
```diff
@@ -401,29 +503,29 @@ def create(self, values):
...
-            doc_data = self._parse_document_url(values['url']).get('values', dict())
-            for key, value in doc_data.items():
-                values.setdefault(key, value)
 
         slide = super(Slide, self).create(values)
 
+        # avoid fetching external metadata when installing the module (i.e. for demo data)
+        # we also support a context key if you don't want to fetch the metadata when creating a slide
+        if any(values.get(url_param) for url_param in ['url', 'video_url', 'document_google_url', 'image_google_url']) \
+           and not self.env.context.get('install_mode', False) \
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:561]: None being default, no need to give explicit "False" to get.

File: addons/website_slides/models/slide_slide.py [hunk 11]
```diff
@@ -401,29 +503,29 @@ def create(self, values):
...
+           and self.env.context.get('website_slides_fetch_metadata', True):
+            slide_metadata = slide._fetch_external_metadata()
+            if slide_metadata:
+                # only update keys that are not set in the incoming values
+                slide.update({key: value for key, value in slide_metadata.items() if key not in values.keys()})
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:566]: Why not calling ``_on_change_url`` manually ? Something like
```
if not self.env.context.get('install_mode') and not
    self.env.context.get('website_slides_skip_fetch_metadata):
    slide._on_change_url()
```

That way we rely on the onchange and a bit less on "call method and then do an update".

Edit: actually you filter out values not already given at create, ok. Then maybe call ``_fetch_external_metadata`` only if we have values to update, to avoid an unnecessary call. If everything is given, hop hop skip it.

File: addons/website_slides/models/slide_slide.py [hunk 12]
```diff
@@ -401,29 +503,29 @@ def create(self, values):
...
-        if values.get('url') and not values.get('document_id'):
-            doc_data = self._parse_document_url(values['url']).get('values', dict())
-            for key, value in doc_data.items():
-                values.setdefault(key, value)
 
         slide = super(Slide, self).create(values)
 
+        # avoid fetching external metadata when installing the module (i.e. for demo data)
+        # we also support a context key if you don't want to fetch the metadata when creating a slide
+        if any(values.get(url_param) for url_param in ['url', 'video_url', 'document_google_url', 'image_google_url']) \
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:560]: While we are doing "onchange-like stuff because of external data", shoudln't we call ``_on_change_document_binary_content`` if not ``completion_time`` is not given in creation values ? That way our create is complete.

File: addons/website_slides/models/slide_slide.py (hunk 8 above)
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py]: Side note: could be great to make create multi-enabled (but not mandatory, if you feel this is going to take too much time, can be done afterwards).

File: addons/website_slides/models/slide_slide.py [hunk 13]
```diff
@@ -433,6 +535,20 @@ def write(self, values):
...
+           and not self.env.context.get('install_mode', False) \
+           and self.env.context.get('website_slides_fetch_metadata', True):
+            slide_metadata = self._fetch_external_metadata()
+            if slide_metadata:
+                # only update keys that are not set in the incoming values and for which we don't have a value yet
+                self.update({
+                    key: value
+                    for key, value in slide_metadata.items()
+                    if key not in values.keys() and not any(slide[key] for slide in self)
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:593]: Same here: we could call _fetch_external_metadata only if some fields are missing. Purpose would be to speedup if possible when we already have some information.

File: addons/website_slides/models/slide_slide.py [hunk 14]
```diff
@@ -433,6 +535,20 @@ def write(self, values):
             self.date_published = datetime.datetime.now()
             self._post_publication()
 
+        # avoid fetching external metadata when installing the module (i.e. for demo data)
+        # we also support a context key if you don't want to fetch the metadata when modifying a slide
+        if any(values.get(url_param) for url_param in ['url', 'video_url', 'document_google_url', 'image_google_url']) \
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:584]: As the last 3 are related, is it necessary ? Or maybe it calls _write and not write...

File: addons/website_slides/models/slide_slide.py [hunk 15]
```diff
@@ -700,165 +816,231 @@ def _compute_quiz_info(self, target_partner, quiz_done=False):
     # Parsing methods
     # --------------------------------------------------
 
-    @api.model
-    def _fetch_data(self, base_url, params, content_type=False):
-        result = {'values': dict()}
+    def _fetch_external_metadata(self, fetch_image=True, return_when_not_found=False):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:863]: return_when_not_found is strangely named. Especially it changes the output of sub-methods ... don't really like it but hey. Maybe a renaming would ease understanding, like ``allow_return_error`` ? ``return_void_on_not_found=True`` ?

Comment (tde-banana-odoo) [reply addons/website_slides/models/slide_slide.py:863]: Counter attack: make sub methods return two things (values, error), and handle only values here and error when needed ? Better than a ping-pong style parameter.

Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:863]: fetch_image -> include_image maybe ? or ``'image_url_only=True`` ?

**Review by tde-banana-odoo — COMMENTED** [review]
Review part 1 (commit 1 + commit 2 until the "fetch external data" methods). Tired for today, I think you already have some stuff to do until I have another coffee .

And by the way: thanks for your work, it is clean :) . Slide model gains a lot in clarity.

File: addons/website_slides/models/slide_slide.py [hunk 16]
```diff
@@ -700,165 +816,231 @@ def _compute_quiz_info(self, target_partner, quiz_done=False):
...
         params = {}
         params['projection'] = 'BASIC'
         if 'google.drive.config' in self.env:
-            access_token = self.env['google.drive.config'].get_access_token()
+            access_token = False
+            try:
+                access_token = self.env['google.drive.config'].get_access_token()
+            except RedirectWarning as e:
+                pass  # ignore and use the 'key' fallback
+            except UserError as e:
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:984]: I think except can be chained (except A, B )

File: addons/website_slides/models/slide_slide.py [hunk 17]
```diff
@@ -700,165 +816,231 @@ def _compute_quiz_info(self, target_partner, quiz_done=False):
...
+                    except Exception:
+                        pass  # fail silently as this is nice to have
+            elif mime_type in sheet_mimetypes:
+                slide_metadata['slide_type'] = 'sheet'
+            elif mime_type in doc_mimetypes:
+                slide_metadata['slide_type'] = 'doc'
+            elif mime_type in slides_mimetypes:
+                slide_metadata['slide_type'] = 'slides'
+            elif mime_type and mime_type.startswith('image/'):
+                # image and videos should be input using another "slide_category" but let's be nice and
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:1076]: Won't we update the slide_category ?

File: addons/website_slides/models/slide_slide.py [hunk 18]
```diff
@@ -872,6 +1054,18 @@ def _default_website_meta(self):
     # Data / Misc
     # ---------------------------------------------------------
 
+    def _get_completion_time_pdf(self, data):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:1169]: b64_data maybe ? Or tell it shoudl be base64 encoded ?

File: addons/website_slides/static/src/js/slides_course_fullscreen_player.js [hunk 19]
```diff
@@ -512,14 +512,22 @@
                     slideData.embedUrl = slideData.embedCode ? scheme + slideData.embedCode + separator + $.param(params) : "";
                 } else if (slideData.category === 'infographic') {
                     slideData.embedUrl = _.str.sprintf('/web/image/slide.slide/%s/image_1024', slideData.id);
-                } else if (_.contains(['document', 'presentation'], slideData.category)) {
+                } else if ( slideData.category === 'document') {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_course_fullscreen_player.js:610]: Isn't it an unnecessary whitespace here ?

File: addons/website_slides/static/src/js/slides_upload.js [hunk 20]
```diff
@@ -99,11 +102,12 @@ var SlideUploadDialog = Dialog.extend({
             });
         }));
     },
-    _fetchUrlPreview: function (url) {
+    _fetchUrlPreview: function (url, slideType) {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:105]: ``slideCategory`` maybe ?

File: addons/website_slides/static/src/js/slides_upload.js [hunk 21]
```diff
@@ -126,46 +130,47 @@ var SlideUploadDialog = Dialog.extend({
      * @private
      */
     _formValidateGetValues: async function (forcePublished) {
-        var canvas = this.$('#data_canvas')[0];
+        var slideType = 'document';
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:136]: ``slideCategory`` ?

File: addons/website_slides/static/src/js/slides_upload.js [hunk 22]
```diff
@@ -126,46 +130,47 @@ var SlideUploadDialog = Dialog.extend({
...
+        } else {
+            sourceType = this.$('input[name="source_type"]:checked').data('value');
+        }
         var values = _.extend({
             'channel_id': this.channelID,
             'name': this._formGetFieldValue('name'),
-            'url': this._formGetFieldValue('url'),
+            'slide_category': slideType,
+            'source_type': sourceType,
+            'document_google_url': this._formGetFieldValue('document_google_url'),
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:153]: Linting: Sockets ordering ?

File: addons/website_slides/static/src/js/slides_upload.js [hunk 23]
```diff
@@ -126,46 +130,47 @@ var SlideUploadDialog = Dialog.extend({
...
-                    'binary_content': fileData,
-                });
-            } else {
-                _.extend(values, {
-                    'image_1920': fileData,
-                });
-            }
+            _.extend(values, {
+                'slide_category': 'infographic',
+                'binary_content': this.file.type === 'image/svg+xml' ? await this._svgToPNG() : this.file.data
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:175]: Linting: ending "," maybe ?

File: addons/website_slides/static/src/js/slides_upload.js [hunk 24]
```diff
@@ -327,10 +332,15 @@ var SlideUploadDialog = Dialog.extend({
...
         this.slide_category_data = {
-            presentation: {
+            document: {
                 icon: 'fa-file-pdf-o',
-                label: _t('Presentation'),
-                template: 'website.slide.upload.modal.presentation',
+                label: _t('Document'),
+                template: 'website.slide.upload.modal.document',
+            },
+            infographic: {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:343]: image maybe, as we now use image ? Note that keeping infographic helps grepping it, thinking of it.

File: addons/website_slides/static/src/js/slides_upload.js [hunk 25]
```diff
@@ -378,6 +388,68 @@ var SlideUploadDialog = Dialog.extend({
...
     },
 
+    /**
+     * When the URL changes for slides of types infographic, document and video, we attempt to fetch
+     * some metadata on YouTube / Google Drive (such as a name, a title, a duration, ...).
+     *
+     * @param {string} url
+     * @param {string} slideType
+     */
+    _updateSlideMetadata: function (url, slideType) {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_upload.js:401]: slideCategory ?

File: addons/website_slides/static/src/scss/website_slides.scss [hunk 26]
```diff
@@ -503,6 +503,25 @@ $o-wslides-fs-side-width: 300px;
...
+        line-height: 1.5rem; // necessary to align label correctly with radio button
+
+    }
+
+    .was-validated .form-check .form-check-label {
+        @include o-w-preserve-base; // avoid ugly green when form is marked as 'was-validated'
+    }
+}
+
+.o_wslides_slide_upload_loading {
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/website_slides/static/src/scss/website_slides.scss:529]: Ok :) let us avoid adding "yet another unrelated cleaning", we can live with it.

File: addons/website_slides/static/src/xml/website_slides_upload.xml [hunk 27]
```diff
@@ -188,4 +291,14 @@
             </form>
         </div>
     </t>
+
+    <!-- Misc -->
+    <t t-name="website.slide.upload.modal.loading">
+        <div class="o_wslides_slide_upload_loading position-absolute h-100 w-100 text-center d-flex flex-column justify-content-center">
+            <h2 class="text-white">
+                <span class="fa fa-spinner fa-pulse"></span>
+                <span>Loading slide information...</span>
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/xml/website_slides_upload.xml:318]: slide ? not "lesson" ? or juste "Loading information" maybe ?

File: addons/website_slides/models/slide_slide.py [hunk 28]
```diff
@@ -433,6 +455,23 @@ def _compute_youtube_id(self):
             else:
                 slide.youtube_id = False
 
+    @api.depends('video_url', 'video_source_type')
+    def _compute_vimeo_id(self):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:459]: Can't we compute video source type and video ID in the same method ? Just to avoid regexing twice ... Not sure it helps, just asking to check if we can reduce a bit computation complexity.

File: addons/website_slides/models/slide_slide.py [hunk 29]
```diff
@@ -1042,6 +1083,72 @@ def _fetch_google_drive_metadata(self, fetch_image, return_when_not_found=False)
 
         return slide_metadata
 
+    def _fetch_vimeo_metadata(self, fetch_image=True, return_when_not_found=False):
```
Comment (tde-banana-odoo) [inline addons/website_slides/models/slide_slide.py:1091]: Same remarks for parameters + return type, if we update other methods, just propagate here :) .

File: addons/website_slides/static/src/js/slides_course_fullscreen_player.js (hunk 5 above)
Comment (tde-banana-odoo) [reply addons/website_slides/static/src/js/slides_course_fullscreen_player.js:142]: Ok ^^

File: addons/website_slides/tests/test_slide_utils.py [hunk 30]
```diff
@@ -212,3 +212,34 @@ def test_video_urls(self):
         })
         self.assertEqual('google_drive', slide.video_source_type)
         self.assertEqual('1qU5nHVNbz_r84P_IS5kDzoCuC1h5ZAZR', slide.google_drive_id)
+
+        # test URL from Vimeo
+        slide = Slide.create({
```
Comment (tde-banana-odoo) [inline addons/website_slides/tests/test_slide_utils.py:217]: Could probably be done in a dedicated unit test (test_video_youtube, test_video_vimeo, ...) .

File: addons/website_slides/tests/test_slide_utils.py [hunk 31]
```diff
@@ -212,3 +212,34 @@ def test_video_urls(self):
...
+            'name': 'dummy',
+            'channel_id': self.channel.id,
+            'video_url': 'https://vimeo.com/545859999',
```
Comment (tde-banana-odoo) [inline addons/website_slides/tests/test_slide_utils.py:220]: Youtube example was great: list of sources, expected, then iter on the list. That way addins "a strange test case" is easy :) .

**Review by tde-banana-odoo — COMMENTED** [review]
Review part 2 ! Finished the diff, looking good. Vimeo integration is quite a small diff in the end, meaning our model is probably not that bad :) . Some comments / linting while reading, not much "maybe we could recode that model" this time I promise.

Cheers !

File: addons/website_slides/models/slide_slide.py (hunk 16 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py]: You have to "tuplelize" them to make it work

`except (A,B)`

Learned something today yay.

File: addons/website_slides/models/slide_slide.py (hunk 17 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:1074]: It would be very hard to support on the website form.
A bit of a side-case really, not sure we should go into complicated handling.

File: addons/website_slides/models/slide_slide.py (hunk 18 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:1168]: Actually not sure here, I think we pass the method bytes, not base64 encoded stuff.

File: addons/website_slides/static/src/js/slides_upload.js (hunk 24 above)
Comment (awa-odoo) [reply addons/website_slides/static/src/js/slides_upload.js:343]: This matches the slide_categories, so it's still "infographic" (the technical value).

File: addons/website_slides/static/src/xml/website_slides_upload.xml (hunk 27 above)
Comment (awa-odoo) [reply addons/website_slides/static/src/xml/website_slides_upload.xml]: Ah ! Luc was on the hunt for those, we replaced a bit everywhere for "content", good catch !

**Review by awa-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 

Fiou, that was a big one !
Addressed everything, most is done as requested but some points we need to fight a bit, I let you see underlying comments.
Still need to re-test the main flows to make sure I did not break anything, but that will be for tomorrow.

Thanks for the in-depth review, I hope we are getting there, can't wait to see this one merged.

File: addons/website_slides/static/src/js/slides_course_fullscreen_player.js [hunk 32]
```diff
@@ -129,6 +130,98 @@
...
+        /**
+         * Every time the video changes position, both while viewing and also when seeking manually,
+         * Vimeo triggers this handy 'timeupdate' event.
+         * We use it to set the slide as completed as soon as we reach the end (30 last seconds).
+         *
+         * See https://developer.vimeo.com/player/sdk/reference#timeupdate for more information
+         *
+         * @param {Object} eventData the 'timeupdate' event data
+         */
+        onVideoTimeUpdate: async function (eventData) {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_course_fullscreen_player.js:205]: ``_onVideoTimeUpdate`` ?

File: addons/website_slides/static/src/js/slides_course_fullscreen_player.js [hunk 33]
```diff
@@ -129,6 +130,98 @@
...
+                }
+            }
+        },
+
+        /**
+         * When the player triggers the 'ended' event, we go to the next slide if there is one.
+         *
+         * See https://developer.vimeo.com/player/sdk/reference#ended for more information
+         */
+        onVideoEnded: function () {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/js/slides_course_fullscreen_player.js:218]: ``_onVideoEnded`` ? + define before _onVideoTimeUpdate ?

**Review by tde-banana-odoo — COMMENTED** [review]
Final check after comments ! Mainly looked at model, as I think we already went into details :) . Some strange stuff with related + create/write overrides, could be great to be sure we cannot cleanup a bit further. Otherwise thanks for your work !

Cheers !

File: addons/website_slides/models/slide_slide.py [hunk 34]
```diff
@@ -315,62 +383,96 @@ def _compute_user_membership_id(self):
...
-                record.embed_code = False
+                slide.youtube_id = False
+
+    @api.depends('url', 'document_google_url', 'image_google_url', 'video_url')
+    def _compute_google_drive_id(self):
+        """ Extracts the Google Drive ID from the url based on the slide category. """
 
-    @api.onchange('url')
+        for slide in self:
+            url = slide.url or slide.document_google_url or slide.image_google_url or slide.video_url
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:487]: I just tested again, definitely does not work, even when "url" is in the view.

File: addons/website_slides/models/slide_slide.py (hunk 6 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:504]: Yep, tried it again with "url" in the view, no changes, nothing triggers when I change the value of my "video_url" field.

File: addons/website_slides/models/slide_slide.py (hunk 7 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:524]: I agree there is something strange with related field and the onchange behavior.
But I don't think there is much I can do about it, except deep dive into the web framework for debugging?

File: addons/website_slides/models/slide_slide.py (hunk 11 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:573]: That would actually be a bit complicated because the fact that "we have all the information" depends on the slide_category.

So I would have to do:
If it's a video and we have title / image / slide_type / duration in values then don't fetch.
If it's an image and we have title / slide_type in values then don't fetch.
etc...
Because if only one of those is missing, then we want to complete the data.

I'm thinking that the context key will be more clear in the end.
Also, I know that it's not a strong argument, but the previous code worked the same way, it fetched the metadata without checking if we have all the information.

If you really think we need to change this, I can give it a try in a separate commit to see how it looks.
Will have to re-test most flows as well.

File: addons/website_slides/models/slide_slide.py (hunk 14 above)
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:594]: I tried again putting "url" in the view and printing "values" and I get:

```{'public_views': 0, 'is_published': False, '__last_update': False, 'name': 'sample video several minutes.mp4', 'is_category': False, 'tag_ids': [[6, False, []]], 'active': True, 'channel_id': 1, 'slide_category': 'video', 'slide_type': 'google_drive_video', 'source_type': 'local_file', 'url': False, 'video_url': 'https://drive.google.com/file/d/1qU5nHVNbz_r84P_IS5kDzoCuC1h5ZAZR/view?usp=sharing', 'user_id': 2, 'completion_time': 0.0718, 'is_preview': False, 'slide_resource_downloadable': False, 'description': '<p><br></p>', 'quiz_first_attempt_reward': 10, 'quiz_second_attempt_reward': 7, 'quiz_third_attempt_reward': 5, 'quiz_fourth_attempt_reward': 2, 'sequence': 6}```

As you can see, it passes url to False.
Maybe I'm missing something but it looks like it doesn't work.

File: addons/website_slides/models/slide_slide.py [hunk 35]
```diff
@@ -124,22 +124,22 @@ class Slide(models.Model):
...
-        ('external', 'External (Google Drive)')],
+        ('local_file', 'Upload from Device'),
+        ('external', 'Retrieve from Google Drive')],
         default='local_file', required=True)
     # generic
     url = fields.Char('External URL', help="URL of the Google Drive file or URL of the YouTube video")
     binary_content = fields.Binary('File', attachment=True)
     slide_resource_ids = fields.One2many('slide.slide.resource', 'slide_id', string="Additional Resource for this slide")
-    slide_resource_downloadable = fields.Boolean('Allow Download', default=True, help="Allow the user to download the content of the slide.")
+    slide_resource_downloadable = fields.Boolean('Allow Download', default=False, help="Allow the user to download the content of the slide.")
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides/models/slide_slide.py:142]: Yes, I have a bullet-point about that in my commit message:

> - "Allow Download" of slide content is now False by default
>   As it was nice to have for "Documentation" courses but does not really make
>   sense for "Training" ones.

Maybe it's not clear enough?

**Review by awa-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 

Thanks for those last comments.
(It allowed me to fix a stacktrace as well in a side-case so yay :eyes: )

I fixed most of them.

The only ones left are the related stuff.
Where I did a few more tests and even when putting "url" in the view, it doesn't work as you would think.
I can do more digging in the web framework but I'm afraid it's a limitation.

There is also the "don't fetch metadata if we have all information" thing.
See underlying response for that, still a bit of a fight going on :fist_right: :fist_left: 

Again, thanks for the reviews & cheers!

**Comment by tde-banana-odoo** [comment]
@awa-odoo Ok for the create / write, we can live with it, let us keep the context key.

For the url field, did you try with a force_save ? Maybe web client does not send anything as it comes from a related or something like that ... or is your related readonly=False ?

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: mainly code move (cleaning / splitting methods fetching external data) - calls have timeout and context keys allow to bypass them - Markup is mainly updated compared to current master - Html is called with something rendered using QWeb on a widget controller by slides.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
